core/test/mpi/base/bindings.cpp

```diff
@@ -0,0 +1,782 @@
+        ASSERT_EQ(recv_array.get_data()[0], 1);
+        ASSERT_EQ(recv_array.get_data()[1], 2);
+        ASSERT_EQ(recv_array.get_data()[2], 3);
+        ASSERT_EQ(recv_array.get_data()[3], 4);
+    }
+}
+
+
+TEST_F(MpiBindings, CanPutValuesWithLockAll)
+{
+    using ValueType = int;
+    using window = gko::mpi::window<ValueType>;
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    int* data;
+    if (my_rank == 0) {
+        data = new ValueType[4]{1, 2, 3, 4};
+    } else {
+        data = new ValueType[4]{0, 0, 0, 0};
+    }
+    auto win = window(data, 4 * sizeof(ValueType), comm);
+    win.lock_all();
+    if (my_rank == 0) {
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                gko::mpi::put<ValueType>(data, 4, rank, 0, 4, win);
+                win.flush(rank);
+            }
+        }
+    }
+    win.unlock_all();
+    gko::mpi::synchronize(comm);
+    ASSERT_EQ(data[0], 1);
+    ASSERT_EQ(data[1], 2);
+    ASSERT_EQ(data[2], 3);
+    ASSERT_EQ(data[3], 4);
+    delete data;
+}
+
+
+TEST_F(MpiBindings, CanPutValuesWithExclusiveLock)
+{
+    using ValueType = int;
+    using window = gko::mpi::window<ValueType>;
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    int* data;
+    if (my_rank == 0) {
+        data = new ValueType[4]{1, 2, 3, 4};
+    } else {
+        data = new ValueType[4]{0, 0, 0, 0};
+    }
+    auto win = window(data, 4 * sizeof(ValueType), comm);
+    if (my_rank == 0) {
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                win.lock(rank, 0, window::lock_type::exclusive);
+                gko::mpi::put<ValueType>(data, 4, rank, 0, 4, win);
+                win.flush(rank);
+                win.unlock(rank);
+            }
+        }
+    }
+    gko::mpi::synchronize(comm);
+    ASSERT_EQ(data[0], 1);
+    ASSERT_EQ(data[1], 2);
+    ASSERT_EQ(data[2], 3);
+    ASSERT_EQ(data[3], 4);
+    delete data;
+}
+
+
+TEST_F(MpiBindings, CanPutValuesWithFence)
+{
+    using ValueType = int;
+    using window = gko::mpi::window<ValueType>;
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    auto send_array = gko::Array<ValueType>{ref};
+    auto recv_array = gko::Array<ValueType>{ref};
+    int* data;
+    if (my_rank == 0) {
+        data = new ValueType[4]{1, 2, 3, 4};
+    } else {
+        data = new ValueType[4]{0, 0, 0, 0};
+    }
+    auto win = window(data, 4 * sizeof(ValueType), comm);
+    win.fence();
+    if (my_rank == 0) {
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                gko::mpi::put<ValueType>(data, 4, rank, 0, 4, win);
+            }
+        }
+    }
+    win.fence();
+    gko::mpi::synchronize(comm);
```

Comment:
If you sync the whole thing at this point, I guess you're not really testing the fence.
```suggestion
```